src/lib.rs

```diff
@@ -260,15 +266,21 @@ impl NixPath for [u8] {
     }
 
     fn with_nix_path<T, F>(&self, f: F) -> Result<T>
-            where F: FnOnce(&CStr) -> T {
-        let mut buf = [0u8; PATH_MAX as usize];
-
+    where
+        F: FnOnce(&CStr) -> T,
+    {
         if self.len() >= PATH_MAX as usize {
-            return Err(Errno::ENAMETOOLONG)
+            return Err(Errno::ENAMETOOLONG);
         }
 
-        buf[..self.len()].copy_from_slice(self);
-        match CStr::from_bytes_with_nul(&buf[..=self.len()]) {
+        let buf = MaybeUninit::<[u8; PATH_MAX as usize]>::uninit().as_mut_ptr() as *mut u8;
```

Comment:
Are you sure this doesn't cause the usual problem with as_ptr methods, where the temporary `MaybeUninit::<[u8; PATH_MAX as usize]>::uninit()` is destroyed at the end of this statement?
I would always declare an explicit `buf` variable and then a separate `buf_ptr`, to make the lifetime of `buf` explicit.

Reply:
Thanks for catching that! Confirmed on godbolt assembly with these two examples:

```rust
#![feature(bench_black_box)]

pub fn main() {
    // Optimizes to just `sub     rsp, 32`. Yikes!
    let buf = std::mem::MaybeUninit::new(String::from("abc")).as_mut_ptr() as *mut u8;
    std::hint::black_box(buf);
}
```

```rust
#![feature(bench_black_box)]

pub fn main() {
    // Performs the allocation
    let mut buf = std::mem::MaybeUninit::new(String::from("abc"));
    let buf = buf.as_mut_ptr() as *mut u8;
    std::hint::black_box(buf);
}
```